### ISSUE TITLE
Methods defined on anonymous classes cause typing errors

### ISSUE THREAD

**Issue by paracycle** [sorbet/sorbet#3609]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0AClass.new%20do%0A%20%20def%20match%3F(request)%0A%20%20%20%2042%0A%20%20end%0Aend%0A%0A%0Afoo%20%3D%20T.let(nil%2C%20T.nilable(String))%0Afoo.match%3F(%22bar%22))

```ruby
# typed: true
Class.new do
  def match?(request)
    42
  end
end


foo = T.let(nil, T.nilable(String))
foo.match?("bar")
```

#### Observed output

```
editor.rb:10: Non-private call to private method Object#match? https://srb.help/7031
    10 |foo.match?("bar")
        ^^^^^^^^^^^^^^^^^
    editor.rb:3: Defined as
     3 |  def match?(request)
          ^^^^^^^^^^^^^^^^^^^
Errors: 1
```

#### Expected behavior

I would have expected `match?` to not have been registered as a private instance method of `Object`, but instead that I got this error:
```
editor.rb:4: Method match? does not exist on NilClass component of T.nilable(String) https://srb.help/7003
     10 |foo.match?("bar")
        ^^^^^^^^^^^^^^^^^
```
- - -

This seems to be related to the changes in https://github.com/sorbet/sorbet/pull/3505 /cc @bss

**Comment by bss** [comment]
@paracycle This is definitely related to #3505. I'll take a look to try and figure out what's going on.

**Comment by bss** [comment]
So, what's going on is that classes defined by `Class.new` is special. There is a rewrite-pass that will rewrite this:
```
A = Class.new do
  def foo; end
end
```
To:
```
class A
  def foo; end
end
```

The bare case:
```
Class.new do
  def bar; end
end
```
Is not rewritten, instead we end up with a symbol table where `#bar` is defined on `Object`:
```
$ ./bazel-bin/main/sorbet -e "Class.new do def foo; end; end; nil.foo" -p symbol-table
👋 Hey there! Heads up that this is not a release build of sorbet.
Release builds are faster and more well-supported by the Sorbet team.
Check out the README to learn how to build Sorbet in release mode.
To forcibly silence this error, either pass --silence-dev-message,
or set SORBET_SILENCE_DEV_MESSAGE=1 in your shell environment.

-e:1: Non-private call to private method Object#foo https://srb.help/7031
     1 |Class.new do def foo; end; end; nil.foo
                                        ^^^^^^^
    -e:1: Defined as
     1 |Class.new do def foo; end; end; nil.foo
                     ^^^^^^^
class ::<root> < ::Object ()
  class ::<Class:<root>>[<AttachedClass>] < ::<Class:Object> ()
    method ::<Class:<root>>#<static-init> (<blk>) @ -e:1
      argument <blk><block> @ Loc {file=-e start=??? end=???}
  class ::Object < ::BasicObject (Kernel) @ https://github.com/sorbet/sorbet/tree/master/rbi/core/object.rbi#L27
    method ::Object#foo : private (<blk>) @ -e:1
      argument <blk><block> @ Loc {file=-e start=??? end=???}
```

I just compiled a version of sorbet locally without the private-call check, and as expected (based on the symbol table above) it unfortunately doesn't error out:
```
$ ./bazel-bin/main/sorbet --typed=true -e "Class.new do def foo; end; end; nil.foo" -p symbol-table
👋 Hey there! Heads up that this is not a release build of sorbet.
Release builds are faster and more well-supported by the Sorbet team.
Check out the README to learn how to build Sorbet in release mode.
To forcibly silence this error, either pass --silence-dev-message,
or set SORBET_SILENCE_DEV_MESSAGE=1 in your shell environment.

class ::<root> < ::Object ()
  class ::<Class:<root>>[<AttachedClass>] < ::<Class:Object> ()
    method ::<Class:<root>>#<static-init> (<blk>) @ -e:1
      argument <blk><block> @ Loc {file=-e start=??? end=???}
  class ::Object < ::BasicObject (Kernel) @ https://github.com/sorbet/sorbet/tree/master/rbi/core/object.rbi#L27
    method ::Object#foo : private (<blk>) @ -e:1
      argument <blk><block> @ Loc {file=-e start=??? end=???}

No errors! Great job.
```

So, the bug is actually that `Class.new` isn't handled correctly. One way to "fix" it could be to not support that style of `Class.new` at all, and essentially skip any class/methods defined that way in the symbol table.

**Comment by paracycle** [comment]
Oh wow, great find @bss, thanks for looking into this. We have some instances of bare anonymous class definitions in our codebase, that is mostly to do with some mock objects in test case. That means, those have been infecting our `Object` method list since the beginning 😲  Your private-call check, thankfully, has surfaced that.

@jez @elliottt Is there a way we can fix this so that bare anonymous class definitions don't end up defining their methods on `Object` (private or not)?

**Comment by jez** [comment]
I think that the solution @bss suggests of getting rid of those nodes from the AST is the only way to support this in Sorbet.

We can't stick the methods on anonymous class definitions in a rewriter pass because all the rewriter passes operate in parallel, so there's no easy way to coordinate to ensure that anonymous class names are globally unqiue.

**Comment by paracycle** [comment]
> ... there's no easy way to coordinate to ensure that anonymous class names are globally unqiue

Do we need them to be globally unique **and** stable? If not, we could just generate some GUIDs during the rewriting process. Not sure about the performance implications of doing that though.

**Comment by jez** [comment]
So actually I thought about it longer and getting unique IDs is actually easy because the pass is parallelized by file, so we could get unique IDs using a scheme like `<FileRef::_id>_<Rewriterer::uniqueCounter++>`

But then the problem becomes that we can't put these class defs inside method defs, because that is a syntax error in Ruby.

https://sorbet.run/#%23%20typed%3A%20true%0Adef%20before%0A%20%20x%20%3D%20Class.new(Integer)%0Aend%0A%0Adef%20after%0A%20%20%23%20AnonymousClass_%3CFileRef%3A%3A_id%3E_%3CuniqueCounter%2B%2B%3E_x%0A%20%20class%20AnonymousClass_1_1_x%0A%20%20end%0A%0A%20%20x%20%3D%20T.let(T.unsafe(Kernel).raise(%22unimplemented%22)%2C%20T.class_of(Integer))%0Aend

**Comment by jez** [comment]
> Do we need them to be globally unique **and** stable?

Yes, everything that ever ends up in the SymbolTable must be stable, otherwise LSP won't work.

**Comment by jez** [comment]
The other thing: almost every that `local_var = Class.new {...}` occurs in a codebase, it's because it couldn't have been written with a normal class def. Specifically, it's relying on the fact that the block passed to `Class.new` has access to local variables that were in scope lexically. In code:

```ruby
x = 0
y = Class.new {p x}
```

is ok, but

```ruby
x = 0
class AnonymousClass_1_1_y
  p x
end
```

is an error, because the local variable `x` is not in scope inside the class def.

Sorbet only has one notion of class definitions—the `class ...; end` one

So the naive approach to fix this would be:

1. convert `x = Class.new` to a class def
2. move that class def to the end of the file (or end of the containing class)

but if that approach works in a rewriter pass, you may as well just do that yourself in the code verbatim.

The reason why people *don't* do that is almost always because it's not possible, and they're acccessing variables from teh closure, which no amount of desugaring will help with.

All of which is to say: Sorbet is somewhat fundamentally incompatible with anonymous classes, and we recommend avoiding them.

There are creative solutions here, for example:

- Rewrite `x = Class.new {...}` to `x = T.let(..., Class)`, and throw away the proc body
- Rewrite `x = Class.new {...}` to `x = T.let(..., Class)`, and rewrite methods inside the proc body from `def foo` to `def <defined_inside_anonymous_class_dont_shadow_things_foo>` so they still show up on Object (or the containing class) but they don't interfere with real things

**Comment by magni-** [comment]
From @alexevanczuk on [the Sorbet slack](https://sorbet-ruby.slack.com/archives/CHN2L03NH/p1734623279948769?thread_ts=1734618908.601069&cid=CHN2L03NH):
> Okay this I think will do the trick:
> ```ruby
> config.before do
>   # In order to avoid test pollution, we can remove the `TestClass` namespace before each test.
>   # This is also helpful if the same `T::Struct` is referenced multiple times, you end up with an error that looks like this:
>   #
>   # ArgumentError:
>   #      Attempted to redefine prop :const_1 on class TestClass::ExamplePermissiveStruct
>   #
>   Object.send(:remove_const, :TestClass) if Object.constants.include?(:TestClass) # rubocop:disable RSpec/RemoveConst,Sorbet/ConstantsFromStrings
>   Object.const_set(:TestClass, Module.new)
> end
>```

**Comment by dgollahon-plaid** [comment]
Given that #10135 was closed as a dupe (i think makes sense) and this issue has been open for 5 years, is there some kind of lightweight workaround that could be done in sorbet? I think if Sorbet just completely ignored defs inside `Class.new` that would be better than misattributing their scope. It was quite confusing to encounter the issue in #10135 because in my case it caused a cascade of thousands of type errors (since the test with the dynamic class was defining `self.new`).

Would it make sense to just skip the def(s)?
